File: src/rest/recommendations-rest-client.ts
import { mapRecommendation, mapReplaySummary } from '../mappers/apple-music-json-mapper';
import type { AppleMusicRestTransport } from './apple-music-rest-transport';
import { mapTopLevelResourceArray } from './rest-json';

/** Recommendations and Replay Apple Music REST. */
export class RecommendationsRestClient {
  constructor(private readonly transport: AppleMusicRestTransport) {}

  async getRecommendations(musicUserToken: string, ids: string[] | null) {
    const query: Record<string, string> = {};
    if (ids?.length) {
      query.ids = ids.join(',');
    }
    const json = await this.transport.getJson('/v1/me/recommendations', query, musicUserToken);
    return mapTopLevelResourceArray(json.data, mapRecommendation);
  }

  async getReplay(musicUserToken: string, year: number | null) {
    const query: Record<string, string> = {};
    if (year != null) {
      query['filter[year]'] = String(year);
    }
    const json = await this.transport.getJson('/v1/me/music-summaries', query, musicUserToken);
    return mapTopLevelResourceArray(json.data, mapReplaySummary);
  }
}

File: android/src/main/java/expo/modules/applemusic/RecommendationsRestClient.kt
package expo.modules.applemusic

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/** Recommendations and Replay Apple Music REST. */
internal class RecommendationsRestClient(
  private val transport: AppleMusicRestTransport,
) {
  suspend fun getRecommendations(musicUserToken: String, ids: List<String>?): List<Map<String, Any?>> =
    withContext(Dispatchers.IO) {
      val query =
        if (!ids.isNullOrEmpty()) {
          mapOf("ids" to ids.joinToString(","))
        } else {
          emptyMap()
        }
      val json = transport.getJson(
          musicUserToken,
          "/v1/me/recommendations", query)
      mapTopLevelResourceArray(json) { AppleMusicJsonMapper.mapRecommendation(it) }
    }

  suspend fun getReplay(musicUserToken: String, year: Int?): List<Map<String, Any?>> =
    withContext(Dispatchers.IO) {
      val query =
        if (year != null) {
          mapOf("filter[year]" to year.toString())
        } else {
          emptyMap()
        }
      val json = transport.getJson(
          musicUserToken,
          "/v1/me/music-summaries", query)
      mapTopLevelResourceArray(json) { AppleMusicJsonMapper.mapReplaySummary(it) }
    }
}
